### PR TITLE
Add swap in YASPool

### PR DESCRIPTION
**Pull request by dpinones**
Closes: #97

### CODE REVIEW

File: src/contracts/yas_pool.cairo [hunk 1]
```diff
@@ -84,25 +185,304 @@ mod YASPool {
...
+                        && sqrt_price_limit_X96 > TickMath::MIN_SQRT_RATIO()
+                } else {
+                    sqrt_price_limit_X96 > slot_0_start.sqrt_price_X96
+                        && sqrt_price_limit_X96 < TickMath::MAX_SQRT_RATIO()
+                },
+                'SPL'
+            );
+
+            unlocked = false;
+            self.unlocked.write(unlocked);
```
Comment (rcatalan98) [inline src/contracts/yas_pool.cairo:229]: Could we use a wrapper function called `lock` and `unlock` something in those lines? It would make everything cleaner.

File: src/interfaces/interface_yas_swap_callback.cairo [hunk 2]
```diff
@@ -0,0 +1,8 @@
+use yas::numbers::signed_integer::i256::i256;
+
+#[starknet::interface]
+trait IYASSwapCallback<TContractState> {
```
Comment (rcatalan98) [inline src/interfaces/interface_yas_swap_callback.cairo:4]:
```suggestion
// In order to do a swap the contract trying to perform this action will have to implement this function in their code.
#[starknet::interface]
trait IYASSwapCallback<TContractState> {
```

File: src/contracts/yas_pool.cairo [hunk 3]
```diff
@@ -47,14 +58,102 @@ mod YASPool {
...
+        // liquidity at the beginning of the swap
+        liquidity_start: u128,
+        // the timestamp of the current block
+        block_timestamp: u64,
+        // the current value of the tick accumulator, computed only if we cross an initialized tick
+        tick_cumulative: i64,
+        // the current value of seconds per liquidity accumulator, computed only if we cross an initialized tick
+        seconds_per_liquidity_cumulative_X128: u256,
+        // whether we've computed and cached the above two accumulators
+        computed_latest_observation: bool
```
Comment (rcatalan98) [inline src/contracts/yas_pool.cairo:96]: Isn't this only used for the Oracle?

File: src/contracts/yas_pool.cairo [hunk 4]
```diff
@@ -84,25 +185,304 @@ mod YASPool {
...
+            zero_for_one: bool,
+            amount_specified: i256,
+            sqrt_price_limit_X96: FixedType,
+            data: Array<felt252>
+        ) -> (i256, i256) {
+            assert(amount_specified.is_non_zero(), 'AS');
+
+            let slot_0_start: Slot0 = self.slot_0.read();
+
+            let mut unlocked = self.unlocked.read();
```
[the remark this answers is not in the document]
Comment (rcatalan98) [reply src/contracts/yas_pool.cairo:215]: We could have internal functions like `check_lock`, `lock`, `unlock`.

File: src/contracts/yas_pool.cairo [hunk 5]
```diff
@@ -84,25 +185,304 @@ mod YASPool {
...
+                if exact_input {
+                    state
+                        .amount_specified_remaining -=
+                            IntegerTrait::<i256>::new(step_amount_in + step_fee_amount, false);
+                    state.amount_calculated = state.amount_calculated
+                        - IntegerTrait::<i256>::new(step_amount_out, false);
+                } else {
+                    state
+                        .amount_specified_remaining +=
+                            IntegerTrait::<i256>::new(step_amount_out, false);
```
Comment (rcatalan98) [inline src/contracts/yas_pool.cairo:317]: Would it make sense to have implemented `Into` from u256 to i256?

File: src/contracts/yas_pool.cairo [hunk 6]
```diff
@@ -84,25 +185,304 @@ mod YASPool {
...
+                        .fee_growth_global_X128 +=
+                            FullMath::mul_div(step_fee_amount, Q128, step_fee_amount);
+                };
+
+                // shift tick if we reached the next price
+                if state.sqrt_price_X96 == step_sqrt_price_next_X96 {
+                    // if the tick is initialized, run the tick transition
+                    if step_initialized {
+                        // crosses an initialized tick
+                        let mut liquidity_net: i128 = Tick::TickImpl::cross(
```
Comment (rcatalan98) [inline src/contracts/yas_pool.cairo:341]: In Uniswap they have a function returning a new value of `cache.tickCumulative` by calling `observeSingle`. Considering we are not using this function but we do need `tick_cumulative` should we consider modifying this variable anyway?

You can check the line [here](https://github.com/Uniswap/v3-core/blob/main/contracts/UniswapV3Pool.sol#L699)

File: src/contracts/yas_pool.cairo [hunk 7]
```diff
@@ -84,25 +185,304 @@ mod YASPool {
...
+            ref self: ContractState,
+            recipient: ContractAddress,
+            zero_for_one: bool,
+            amount_specified: i256,
+            sqrt_price_limit_X96: FixedType,
+            data: Array<felt252>
+        ) -> (i256, i256) {
+            assert(amount_specified.is_non_zero(), 'AS');
+
+            let slot_0_start: Slot0 = self.slot_0.read();
```
Comment (dubzn) [inline src/contracts/yas_pool.cairo:213]:
```suggestion
            let slot_0_start = self.slot_0.read();
```

File: src/contracts/yas_pool.cairo [hunk 8]
```diff
@@ -84,25 +185,304 @@ mod YASPool {
...
+                        amount_1,
+                        sqrt_price_X96: state.sqrt_price_X96,
+                        liquidity: state.liquidity,
+                        tick: state.tick
+                    }
+                );
+
+            self.unlocked.write(true);
+
+            (amount_0, amount_0)
```
Comment (dubzn) [inline src/contracts/yas_pool.cairo:464]: (amount_0, amount_1) ?

File: src/contracts/yas_pool.cairo [hunk 9]
```diff
@@ -84,25 +185,304 @@ mod YASPool {
...
+                    let delta: u256 = step_fee_amount / cache.fee_protocol.into();
+                    step_fee_amount -= delta;
+                    state.protocol_fee += delta.try_into().unwrap();
+                };
+
+                // update global fee tracker
+                if state.liquidity > 0 {
+                    state
+                        .fee_growth_global_X128 +=
+                            FullMath::mul_div(step_fee_amount, Q128, step_fee_amount);
```
Comment (dubzn) [inline src/contracts/yas_pool.cairo:333]: There might be an error here, specifically in the third parameter.
```
// update global fee tracker
if (state.liquidity > 0)
    state.feeGrowthGlobalX128 += FullMath.mulDiv(step.feeAmount, FixedPoint128.Q128, state.liquidity);
```
```suggestion
                            FullMath::mul_div(step_fee_amount, Q128, state.liquidity);
```

File: src/contracts/yas_pool.cairo (hunk 6 above)
Comment (dubzn) [inline src/contracts/yas_pool.cairo:341]:
```suggestion
                        let mut liquidity_net = Tick::TickImpl::cross(
```

File: src/contracts/yas_pool.cairo [hunk 10]
```diff
@@ -84,25 +185,304 @@ mod YASPool {
...
+                            },
+                            cache.seconds_per_liquidity_cumulative_X128, // TODO:
+                            cache.tick_cumulative,
+                            cache.block_timestamp
+                        );
+
+                        // if we're moving leftward, we interpret liquidityNet as the opposite sign
+                        // safe because liquidityNet cannot be type(int128).min
+                        if zero_for_one {
+                            liquidity_net = -liquidity_net;
```
Comment (dubzn) [inline src/contracts/yas_pool.cairo:362]: Are we sure that using `- liquidity_net` is sufficient to invert the sign of `liquidity_net`? There is a method that specifically does this; could you check it? It's called `i128_neg()`

File: src/contracts/yas_pool.cairo [hunk 11]
```diff
@@ -84,25 +185,304 @@ mod YASPool {
...
+                };
+            };
+
+            // update tick and write an oracle entry if the tick change
+            let mut slot_0 = self.slot_0.read();
+            if state.tick != slot_0_start.tick {
+                slot_0.tick = state.tick;
+            }
+            slot_0.sqrt_price_X96 = state.sqrt_price_X96;
+            self.slot_0.write(slot_0);
```
Comment (dubzn) [inline src/contracts/yas_pool.cairo:387]: Could there be an error? The assignment of the `slot_0.sqrt_price_X96` is done within an "else" block.
 ```
// update tick and write an oracle entry if the tick change
        if (state.tick != slot0Start.tick) {
            (uint16 observationIndex, uint16 observationCardinality) =
                observations.write(
                    slot0Start.observationIndex,
                    cache.blockTimestamp,
                    slot0Start.tick,
                    cache.liquidityStart,
                    slot0Start.observationCardinality,
                    slot0Start.observationCardinalityNext
                );
            (slot0.sqrtPriceX96, slot0.tick, slot0.observationIndex, slot0.observationCardinality) = (
                state.sqrtPriceX96,
                state.tick,
                observationIndex,
                observationCardinality
            );
        } else {
            // otherwise just update the price
            slot0.sqrtPriceX96 = state.sqrtPriceX96;
        }
```
```suggestion
            if state.tick != slot_0_start.tick {
                slot_0.tick = state.tick;
            } else {
                 slot_0.sqrt_price_X96 = state.sqrt_price_X96;
            }
```

File: src/utils/utils.cairo [hunk 12]
```diff
@@ -41,7 +41,6 @@ impl Slot0PartialEq of PartialEq<Slot0> {
         (*lhs.sqrt_price_X96) == (*rhs.sqrt_price_X96)
             && (*lhs.tick) == (*rhs.tick)
             && lhs.fee_protocol == rhs.fee_protocol
-            && lhs.unlocked == rhs.unlocked
```
Comment (dubzn) [inline src/utils/utils.cairo:44]: Why did we remove this?

File: src/libraries/tick_math.cairo [hunk 13]
```diff
@@ -22,9 +22,14 @@ mod TickMath {
...
-    const MIN_SQRT_RATIO: u256 = 4295128739;
+    fn MIN_SQRT_RATIO() -> FixedType {
+        FixedTrait::new(4295128739, false)
+    }
+
     /// The maximum value that can be returned from `get_sqrt_ratio_at_tick`. Equivalent to get_sqrt_ratio_at_tick(MAX_TICK).
-    const MAX_SQRT_RATIO: u256 = 1461446703485210103287273052203988822378723970342;
+    fn MAX_SQRT_RATIO() -> FixedType {
+        FixedTrait::new(1461446703485210103287273052203988822378723970342, false)
+    }
```
Comment (dubzn) [inline src/libraries/tick_math.cairo:32]: Are we sure they represent the same thing? Why was this change necessary? Should we add something like MIN/MAX_SQRT_RATIO_X96 for clarity?

Comment (rcatalan98) [reply src/libraries/tick_math.cairo:32]: I agree this could be useful specially for the future when we will do a refactor. We still need to add the 128.128 type.

File: src/contracts/yas_pool.cairo (hunk 4 above)
[the remark this answers is not in the document]
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: ✅

File: src/contracts/yas_pool.cairo (hunk 1 above)
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: ✅

File: src/contracts/yas_pool.cairo (hunk 7 above)
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: ✅

File: src/contracts/yas_pool.cairo [hunk 14]
```diff
@@ -84,25 +185,304 @@ mod YASPool {
...
+                    slot_0_start.fee_protocol % 16
+                } else {
+                    slot_0_start.fee_protocol.shr(4)
+                },
+                seconds_per_liquidity_cumulative_X128: 0,
+                tick_cumulative: Zeroable::zero(),
+                computed_latest_observation: false
+            };
+
+            let exact_input: bool = amount_specified > Zeroable::zero();
```
[the remark this answers is not in the document]
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: ✅

File: src/contracts/yas_pool.cairo (hunk 6 above)
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: ✅

File: src/contracts/yas_pool.cairo (hunk 11 above)
Comment (dpinones) [reply src/contracts/yas_pool.cairo:383]: In the `if` statement, the assignment of `slot_0.sqrt_price_X96` is also performed. 
```
// update tick and write an oracle entry if the tick change
if (state.tick != slot0Start.tick) {
   ...
   (slot0.sqrtPriceX96, ...) = (state.sqrtPriceX96, ...);
} else {
   slot0.sqrtPriceX96 = state.sqrtPriceX96;
}
```
The code is equivalent

File: src/contracts/yas_pool.cairo (hunk 9 above)
Comment (dpinones) [reply src/contracts/yas_pool.cairo:329]: You're right, good catch!

File: src/utils/utils.cairo (hunk 12 above)
Comment (dpinones) [reply src/utils/utils.cairo:44]: I removed the `unlocked` variable from here because I also removed it from the `Slot0` struct. This variable will be constantly changing, and it didn't seem right to me to have to update the entire struct just for one field. That's why I made it a storage variable. What do you think? @rcatalan98

File: src/interfaces/interface_ERC20.cairo
[the remark this answers is not in the document]
Comment (dpinones) [reply src/interfaces/interface_ERC20.cairo:1]: ✅

File: src/libraries/tick_math.cairo (hunk 13 above)
Comment (dpinones) [reply src/libraries/tick_math.cairo]: ✅

File: src/contracts/yas_pool.cairo (hunk 8 above)
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: ✅

File: src/contracts/yas_pool.cairo (hunk 9 above)
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: ✅

File: src/contracts/yas_pool.cairo (hunk 5 above)
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: ✅

File: src/contracts/yas_pool.cairo (hunk 3 above)
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: ✅

File: src/interfaces/interface_ERC20.cairo
[the remark this answers is not in the document]
Comment (dpinones) [reply src/interfaces/interface_ERC20.cairo:1]: You're right, in the interface, only the functions we use remain

File: src/contracts/yas_pool.cairo (hunk 3 above)
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: I removed these fields since they are only used by the oracle

File: src/contracts/yas_pool.cairo (hunk 6 above)
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: I removed the `tick_cumulative` variable since it is only used by the oracle

File: src/contracts/yas_pool.cairo [hunk 15]
```diff
@@ -84,25 +185,304 @@ mod YASPool {
...
+                if state.amount_specified_remaining.is_non_zero()
+                    && state.sqrt_price_X96 != sqrt_price_limit_X96 {
+                    break;
+                }
+
+                let step_sqrt_price_start_X96 = state.sqrt_price_X96;
+
+                // TODO: test mut
+                let (mut step_tick_next, step_initialized) =
+                    TickBitmap::TickBitmapImpl::next_initialized_tick_within_one_word(
```
[the remark this answers is not in the document]
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: ✅

File: src/contracts/yas_pool.cairo [hunk 16]
```diff
@@ -84,25 +185,304 @@ mod YASPool {
...
+            self.unlocked.write(unlocked);
+
+            let cache = SwapCache {
+                liquidity_start: self.liquidity.read(),
+                block_timestamp: get_block_timestamp(),
+                fee_protocol: if zero_for_one {
+                    slot_0_start.fee_protocol % 16
+                } else {
+                    slot_0_start.fee_protocol.shr(4)
+                },
```
[the remark this answers is not in the document]
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: ✅

File: src/contracts/yas_pool.cairo (hunk 10 above)
Comment (dpinones) [reply src/contracts/yas_pool.cairo:331]: Yes, using the `-` sign is sufficient to invert the sign. In fact, under the hood, it uses the function you mentioned [i128_neg](https://github.com/lambdaclass/yet-another-swap/blob/a396506ab545ab32317ee7f4ccfc51696b3f40d3/src/numbers/signed_integer/i128.cairo#L425)

File: src/contracts/yas_pool.cairo [hunk 17]
```diff
@@ -1,40 +1,54 @@
...
-        // the current price
-        sqrt_price_X96: FixedType,
-        // the current tick
-        tick: i32,
-        // represented as an integer denominator (1/x)%
-        fee_protocol: u8,
-        // whether the pool is locked
-        unlocked: bool
-    }
+    const Q128: u256 = 0x100000000000000000000000000000000;
```
[the remark this answers is not in the document]
Comment (dpinones) [reply src/contracts/yas_pool.cairo]: ✅
